`ImgPlus` images with a new `ROITree` that contains a `RealPointCollection` can't be displayed

I’m in the process of creating better Python ↔ ImageJ/ImgLib2 ROI support for PyImageJ. While working on this, I discovered that `net.imagej.ImgPlus` images that have a new `net.imagej.roi.DefaultROITree` that contains a `net.imglib2.roi.geom.real.DefaultWritableRealPointCollection` of `net.imglib2.RealPoint` points can’t be displayed. Here is a minimal example that replicates this bug:

```java
import net.imagej.ImageJ;
import net.imagej.ImgPlus;
import net.imglib2.RealLocalizable;
import net.imglib2.RealPoint;
import net.imglib2.roi.geom.real.DefaultWritableRealPointCollection;
import net.imglib2.roi.geom.real.RealPointCollection;
import net.imglib2.img.Img;
import net.imagej.roi.DefaultROITree;

import java.io.IOException;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class RealPointCollectionTest {
    private static void run() throws IOException {
        ImageJ ij = new ImageJ();

        // open blobs
        Img img = (Img) ij.io().open(Object.class.getResource("/blobs.png").getPath());
        ImgPlus imgPlus = new ImgPlus(img);

        // create points
        List<RealLocalizable> points =  new ArrayList<>(5);
        points.add(new RealPoint(new double[] {50, 200}));
        points.add(new RealPoint(new double[] {180, 90}));
        points.add(new RealPoint(new double[] {62, 76}));
        points.add(new RealPoint(new double[] {90, 110}));
        points.add(new RealPoint(new double[] {12, 80}));
        RealPointCollection< ? > rpc = new DefaultWritableRealPointCollection<>(points);
        
        // create ROITree and add points
        final DefaultROITree tree = new DefaultROITree();
        tree.addROIs(Arrays.asList(rpc));

        // create metadata map
        Map<String, DefaultROITree> metadata = new HashMap<>();
        metadata.put("rois", tree);

        // attach metadata to the imgplus
        imgPlus.getProperties().putAll(metadata);

        // show blobs
        ij.ui().show(imgPlus);
    }

    public static void main(String...args) throws IOException {
        run();
    }
}
```

And here's the stack trace:

```java
[ERROR] java.lang.reflect.InvocationTargetException
        at java.awt.EventQueue.invokeAndWait(EventQueue.java:1349)
        at java.awt.EventQueue.invokeAndWait(EventQueue.java:1324)
        at org.scijava.thread.DefaultThreadService.invoke(DefaultThreadService.java:115)
        at net.imagej.legacy.ui.LegacyUI.show(LegacyUI.java:191)
        at org.scijava.ui.AbstractUserInterface.show(AbstractUserInterface.java:103)
        at org.scijava.ui.UserInterface.show(UserInterface.java:75)
        at org.scijava.ui.DefaultUIService.show(DefaultUIService.java:247)
        at howto.images.RealPointCollectionTest.run(RealPointCollectionTest.java:48)
        at howto.images.RealPointCollectionTest.main(RealPointCollectionTest.java:52)
Caused by: java.lang.ClassCastException: net.imglib2.RealPoint cannot be cast to net.imglib2.roi.util.RealLocalizableRealPositionable
        at net.imagej.legacy.convert.roi.point.RealPointCollectionWrapper.getCoors(RealPointCollectionWrapper.java:111)
        at net.imagej.legacy.convert.roi.point.RealPointCollectionWrapper.<init>(RealPointCollectionWrapper.java:60)
        at net.imagej.legacy.convert.roi.point.WritableRealPointCollectionToPointRoiConverter.convert(WritableRealPointCollectionToPointRoiConverter.java:70)
        at net.imagej.legacy.convert.roi.point.WritableRealPointCollectionToPointRoiConverter.convert(WritableRealPointCollectionToPointRoiConverter.java:48)
        at net.imagej.legacy.convert.roi.AbstractMaskPredicateToRoiConverter.convert(AbstractMaskPredicateToRoiConverter.java:71)
        at org.scijava.convert.ConvertService.convert(ConvertService.java:68)
        at net.imagej.legacy.convert.ROITreeToOverlayConverter.addROIs(ROITreeToOverlayConverter.java:83)
        at net.imagej.legacy.convert.ROITreeToOverlayConverter.addROIs(ROITreeToOverlayConverter.java:89)
        at net.imagej.legacy.convert.ROITreeToOverlayConverter.convert(ROITreeToOverlayConverter.java:77)
        at org.scijava.convert.ConvertService.convert(ConvertService.java:68)
        at net.imagej.legacy.LegacyImageMap.synchronizeAttachmentsToImagePlus(LegacyImageMap.java:536)
        at net.imagej.legacy.LegacyImageMap.synchronizeAttachmentsToImagePlus(LegacyImageMap.java:512)
        at net.imagej.legacy.LegacyImageMap.registerDisplay(LegacyImageMap.java:268)
        at net.imagej.legacy.LegacyImageMap.registerDisplay(LegacyImageMap.java:252)
        at net.imagej.legacy.display.LegacyImageDisplayViewer.view(LegacyImageDisplayViewer.java:138)
        at net.imagej.legacy.display.LegacyImageDisplayViewer.view(LegacyImageDisplayViewer.java:107)
        at net.imagej.legacy.ui.LegacyUI$1.run(LegacyUI.java:195)
        at org.scijava.thread.DefaultThreadService.lambda$wrap$1(DefaultThreadService.java:211)
        at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:301)
        at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
        at java.awt.EventQueue.access$500(EventQueue.java:97)
        at java.awt.EventQueue$3.run(EventQueue.java:709)
        at java.awt.EventQueue$3.run(EventQueue.java:703)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
        at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
        at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
        at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
        at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
        at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
        at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
        at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

--- a/src/main/java/net/imagej/legacy/convert/roi/point/RealPointCollectionWrapper.java
+++ b/src/main/java/net/imagej/legacy/convert/roi/point/RealPointCollectionWrapper.java
@@ -32,6 +32,7 @@ package net.imagej.legacy.convert.roi.point;
 import ij.gui.PointRoi;
 
 import java.util.Iterator;
+import java.util.function.Supplier;
 
 import net.imagej.legacy.convert.roi.MaskPredicateWrapper;
 import net.imglib2.RealLocalizable;
@@ -48,17 +49,38 @@ import org.scijava.util.FloatArray;
  *
  * @author Alison Walter
  */
-public final class RealPointCollectionWrapper extends PointRoi implements
-	MaskPredicateWrapper<WritableRealPointCollection<RealLocalizableRealPositionable>>
+public final class RealPointCollectionWrapper<L extends RealLocalizable & RealPositionable>
+	extends PointRoi implements
+	MaskPredicateWrapper<WritableRealPointCollection<L>>
 {
 
-	private final WritableRealPointCollection<RealLocalizableRealPositionable> rpc;
+	private final WritableRealPointCollection<L> rpc;
+	private final Supplier<L> newPointCreator;
 	private int numPoints;
 
-	public RealPointCollectionWrapper(
-		final WritableRealPointCollection<RealLocalizableRealPositionable> p)
+	/**
+	 * @deprecated because adding new points through the wrapper is prone to
+	 *             {@link ClassCastException}s. Use
+	 *             {@link #RealPointCollectionWrapper(WritableRealPointCollection, Supplier)}
+	 *             instead.
+	 * @param p the {@link WritableRealPointCollection} that will be wrapped
+	 */
+	@Deprecated
+	@SuppressWarnings("unchecked")
+	public RealPointCollectionWrapper(final WritableRealPointCollection<L> p) {
+		this(p, () -> (L) new RealPoint(2));
+	}
+
+	/**
+	 * @param p the {@link WritableRealPointCollection} that will be wrapped
+	 * @param ptCreator defines how new points will be created when added to the
+	 *          wrapper.
+	 */
+	public RealPointCollectionWrapper(final WritableRealPointCollection<L> p,
+		final Supplier<L> ptCreator)
 	{
 		super(getCoors(p, 0), getCoors(p, 1), countPoints(p));
+		newPointCreator = ptCreator;
 		numPoints = getFloatPolygon().npoints;
 		rpc = p;
 	}
@@ -66,9 +88,7 @@ public final class RealPointCollectionWrapper extends PointRoi implements
 	// -- MaskPredicateWrapper methods --
 
 	@Override
-	public WritableRealPointCollection<RealLocalizableRealPositionable>
-		getSource()
-	{
+	public WritableRealPointCollection<L> getSource() {
 		return rpc;
 	}
 
@@ -77,8 +97,7 @@ public final class RealPointCollectionWrapper extends PointRoi implements
 		// Check if points were added
 		if (getNCoordinates() > numPoints) {
 			while (getNCoordinates() != numPoints) {
-				rpc.addPoint(new RealLocalizableRealPositionableWrapper<>(new RealPoint(
-					2)));
+				rpc.addPoint(newPointCreator.get());
 				numPoints++;
 			}
 		}
@@ -91,8 +110,7 @@ public final class RealPointCollectionWrapper extends PointRoi implements
 		}
 
 		// Update point locations
-		final Iterator<RealLocalizableRealPositionable> itr = rpc.points()
-			.iterator();
+		final Iterator<L> itr = rpc.points().iterator();
 		final float[] xCoor = getContainedFloatPoints().xpoints;
 		final float[] yCoor = getContainedFloatPoints().ypoints;
 		for (int i = 0; i < numPoints; i++)
@@ -101,13 +119,11 @@ public final class RealPointCollectionWrapper extends PointRoi implements
 
 	// -- Helper methods --
 
-	private static <L extends RealLocalizable & RealPositionable> float[] getCoors(
-		final WritableRealPointCollection<L> rpc,
-		final int d)
+	private static <L extends RealLocalizable & RealPositionable> float[]
+		getCoors(final WritableRealPointCollection<L> rpc, final int d)
 	{
 		final FloatArray coor = new FloatArray();
-		final Iterator<L> itr = rpc.points()
-			.iterator();
+		final Iterator<L> itr = rpc.points().iterator();
 		while (itr.hasNext())
 			coor.addValue(itr.next().getFloatPosition(d));
 
@@ -117,8 +133,7 @@ public final class RealPointCollectionWrapper extends PointRoi implements
 	private static <L extends RealLocalizable & RealPositionable> int countPoints(
 		final WritableRealPointCollection<L> rpc)
 	{
-		final Iterator<L> itr = rpc.points()
-			.iterator();
+		final Iterator<L> itr = rpc.points().iterator();
 		int count = 0;
 		while (itr.hasNext()) {
 			itr.next();

--- a/src/main/java/net/imagej/legacy/convert/roi/point/RealPointCollectionWrapper.java
+++ b/src/main/java/net/imagej/legacy/convert/roi/point/RealPointCollectionWrapper.java
@@ -36,6 +36,7 @@ import java.util.Iterator;
 import net.imagej.legacy.convert.roi.MaskPredicateWrapper;
 import net.imglib2.RealLocalizable;
 import net.imglib2.RealPoint;
+import net.imglib2.RealPositionable;
 import net.imglib2.roi.geom.real.WritableRealPointCollection;
 import net.imglib2.roi.util.RealLocalizableRealPositionable;
 import net.imglib2.roi.util.RealLocalizableRealPositionableWrapper;
@@ -100,12 +101,12 @@ public final class RealPointCollectionWrapper extends PointRoi implements
 
 	// -- Helper methods --
 
-	private static <L extends RealLocalizable> float[] getCoors(
-		final WritableRealPointCollection<RealLocalizableRealPositionable> rpc,
+	private static <L extends RealLocalizable & RealPositionable> float[] getCoors(
+		final WritableRealPointCollection<L> rpc,
 		final int d)
 	{
 		final FloatArray coor = new FloatArray();
-		final Iterator<RealLocalizableRealPositionable> itr = rpc.points()
+		final Iterator<L> itr = rpc.points()
 			.iterator();
 		while (itr.hasNext())
 			coor.addValue(itr.next().getFloatPosition(d));
@@ -113,10 +114,10 @@ public final class RealPointCollectionWrapper extends PointRoi implements
 		return coor.getArray();
 	}
 
-	private static <L extends RealLocalizable> int countPoints(
-		final WritableRealPointCollection<RealLocalizableRealPositionable> rpc)
+	private static <L extends RealLocalizable & RealPositionable> int countPoints(
+		final WritableRealPointCollection<L> rpc)
 	{
-		final Iterator<RealLocalizableRealPositionable> itr = rpc.points()
+		final Iterator<L> itr = rpc.points()
 			.iterator();
 		int count = 0;
 		while (itr.hasNext()) {

--- a/src/main/java/net/imagej/legacy/convert/roi/point/WritableRealPointCollectionToPointRoiConverter.java
+++ b/src/main/java/net/imagej/legacy/convert/roi/point/WritableRealPointCollectionToPointRoiConverter.java
@@ -32,11 +32,17 @@ package net.imagej.legacy.convert.roi.point;
 import ij.gui.PointRoi;
 
 import net.imagej.legacy.convert.roi.AbstractMaskPredicateToRoiConverter;
+import net.imglib2.RealLocalizable;
+import net.imglib2.RealPositionable;
 import net.imglib2.roi.geom.real.WritableRealPointCollection;
-import net.imglib2.roi.util.RealLocalizableRealPositionable;
 
+import org.scijava.convert.ConvertService;
 import org.scijava.convert.Converter;
+import org.scijava.plugin.Parameter;
 import org.scijava.plugin.Plugin;
+
+import java.lang.reflect.Type;
+import java.util.function.Supplier;
 
 /**
  * Converts a {@link WritableRealPointCollection} to a {@link PointRoi}. This
@@ -46,9 +52,13 @@ import org.scijava.plugin.Plugin;
  * @author Alison Walter
  */
 @Plugin(type = Converter.class)
-public class WritableRealPointCollectionToPointRoiConverter extends
-	AbstractMaskPredicateToRoiConverter<WritableRealPointCollection<RealLocalizableRealPositionable>, PointRoi>
+public class WritableRealPointCollectionToPointRoiConverter<L extends RealLocalizable & RealPositionable>
+	extends
+	AbstractMaskPredicateToRoiConverter<WritableRealPointCollection<L>, PointRoi>
 {
+
+	@Parameter
+	public ConvertService convert;
 
 	@Override
 	public Class<PointRoi> getOutputType() {
@@ -57,17 +67,18 @@ public class WritableRealPointCollectionToPointRoiConverter extends
 
 	@Override
 	@SuppressWarnings({ "unchecked", "rawtypes" })
-	public Class<WritableRealPointCollection<RealLocalizableRealPositionable>>
-		getInputType()
-	{
+	public Class<WritableRealPointCollection<L>> getInputType() {
 		return (Class) WritableRealPointCollection.class;
 	}
 
 	@Override
-	public PointRoi convert(
-		final WritableRealPointCollection<RealLocalizableRealPositionable> mask)
-	{
-		return new RealPointCollectionWrapper(mask);
+	public PointRoi convert(final WritableRealPointCollection<L> mask) {
+		float[] srcArray = { 0f, 0f };
+		Class<?> ptClass = mask.points().iterator().next().getClass();
+		Converter<float[], L> c = (Converter<float[], L>) convert.getHandler(
+			srcArray, ptClass);
+		Supplier<L> ptCreator = () -> (L) c.convert(srcArray, ptClass);
+		return new RealPointCollectionWrapper<>(mask, ptCreator);
 	}
 
 	@Override

--- a/src/test/java/net/imagej/legacy/convert/roi/point/PointRoiConversionTest.java
+++ b/src/test/java/net/imagej/legacy/convert/roi/point/PointRoiConversionTest.java
@@ -71,7 +71,7 @@ import org.scijava.convert.Converter;
 public class PointRoiConversionTest {
 
 	private PointRoi point;
-	private WritableRealPointCollection<RealLocalizableRealPositionable> rpc;
+	private WritableRealPointCollection<RealPoint> rpc;
 	private WritableRealPointCollection<RealLocalizable> pointRoiWrap;
 	private PointRoi rpcWrap;
 	private ConvertService convertService;
@@ -80,16 +80,13 @@ public class PointRoiConversionTest {
 	public void setup() {
 		point = new PointRoi(new float[] { 12.125f, 17, 1 }, new float[] { -4, 6.5f,
 			30 });
-		final List<RealLocalizableRealPositionable> c = new ArrayList<>();
-		c.add(new RealLocalizableRealPositionableWrapper<>(new RealPoint(
-			new double[] { 12.125, -4 })));
-		c.add(new RealLocalizableRealPositionableWrapper<>(new RealPoint(
-			new double[] { 17, 6.5 })));
-		c.add(new RealLocalizableRealPositionableWrapper<>(new RealPoint(
-			new double[] { 1, 30 })));
+		final List<RealPoint> c = new ArrayList<>();
+		c.add(new RealPoint(12.125, -4));
+		c.add(new RealPoint(17, 6.5 ));
+		c.add(new RealPoint(1, 30));
 		rpc = new DefaultWritableRealPointCollection<>(c);
 		pointRoiWrap = new PointRoiWrapper(point);
-		rpcWrap = new RealPointCollectionWrapper(rpc);
+		rpcWrap = new RealPointCollectionWrapper<>(rpc, () -> new RealPoint(2));
 
 		final Context context = new Context(ConvertService.class);
 		convertService = context.service(ConvertService.class);
@@ -199,8 +196,7 @@ public class PointRoiConversionTest {
 
 		final float[] xp = p.getContainedFloatPoints().xpoints;
 		final float[] yp = p.getContainedFloatPoints().ypoints;
-		final Iterator<RealLocalizableRealPositionable> points = rpc.points()
-			.iterator();
+		final Iterator<RealPoint> points = rpc.points().iterator();
 		int count = 0;
 
 		while (points.hasNext()) {
@@ -215,6 +211,16 @@ public class PointRoiConversionTest {
 		assertEquals(rpc.realMin(1), p.getYBase(), 0);
 		assertEquals(rpc.realMax(0), p.getXBase() + p.getFloatWidth(), 0);
 		assertEquals(rpc.realMax(1), p.getYBase() + p.getFloatHeight(), 0);
+
+		// Add a new point in the wrapper and check for it in the wrapped object
+		p.addPoint(1.0, 1.0);
+		assertEquals(rpc.size(), 3);
+		((RealPointCollectionWrapper<?>) p).synchronize();
+		assertEquals(rpc.size(), 4);
+		Iterator<RealPoint> itr = rpc.points().iterator();
+		for (int i = 0; i < 3; i++)
+			itr.next();
+		assertEquals(new RealPoint(1, 1), itr.next());
 	}
 
 	@Test

--- a/src/test/java/net/imagej/legacy/convert/roi/point/PointRoiWrapperTest.java
+++ b/src/test/java/net/imagej/legacy/convert/roi/point/PointRoiWrapperTest.java
@@ -29,28 +29,31 @@
 
 package net.imagej.legacy.convert.roi.point;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertTrue;
-
 import ij.IJ;
 import ij.ImagePlus;
 import ij.gui.PointRoi;
 
-import java.util.ArrayList;
+import java.util.*;
 import java.util.Iterator;
-import java.util.List;
 
-import net.imglib2.RealLocalizable;
-import net.imglib2.RealPoint;
+import net.imglib2.*;
+import net.imglib2.RandomAccess;
+import net.imglib2.img.Img;
+import net.imglib2.img.array.ArrayImgs;
+import net.imglib2.interpolation.randomaccess.NearestNeighborInterpolatorFactory;
 import net.imglib2.roi.geom.real.DefaultWritableRealPointCollection;
 import net.imglib2.roi.geom.real.RealPointCollection;
 import net.imglib2.roi.geom.real.WritableRealPointCollection;
 
+import net.imglib2.type.numeric.integer.UnsignedByteType;
+import net.imglib2.view.Views;
+import org.apache.commons.math3.analysis.interpolation.LinearInterpolator;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.ExpectedException;
+
+import static org.junit.Assert.*;
 
 /**
  * Tests {@link PointRoiWrapper}

--- a/src/test/java/net/imagej/legacy/convert/roi/point/RealPointCollectionWrapperTest.java
+++ b/src/test/java/net/imagej/legacy/convert/roi/point/RealPointCollectionWrapperTest.java
@@ -29,49 +29,51 @@
 
 package net.imagej.legacy.convert.roi.point;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertTrue;
-
 import ij.IJ;
 import ij.ImagePlus;
 
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
 
 import net.imglib2.RealLocalizable;
 import net.imglib2.RealPoint;
+import net.imglib2.RealRandomAccess;
+import net.imglib2.RealRandomAccessible;
+import net.imglib2.img.array.ArrayImgs;
+import net.imglib2.interpolation.randomaccess.NearestNeighborInterpolatorFactory;
 import net.imglib2.roi.geom.real.DefaultWritableRealPointCollection;
 import net.imglib2.roi.geom.real.WritableRealPointCollection;
-import net.imglib2.roi.util.RealLocalizableRealPositionable;
-import net.imglib2.roi.util.RealLocalizableRealPositionableWrapper;
 
+import net.imglib2.type.numeric.integer.UnsignedByteType;
+import net.imglib2.view.Views;
 import org.junit.Before;
 import org.junit.Test;
+
+import static org.junit.Assert.*;
+import static org.junit.Assert.assertArrayEquals;
 
 /**
  * Tests {@link RealPointCollectionWrapper}
  *
  * @author Alison Walter
+ * @author Gabriel Selzer
  */
 public class RealPointCollectionWrapperTest {
 
-	private WritableRealPointCollection<RealLocalizableRealPositionable> rpc;
-	private RealPointCollectionWrapper wrap;
+	private WritableRealPointCollection<RealPoint> rpc;
+	private RealPointCollectionWrapper<RealPoint> wrap;
 
 	@Before
 	public void setup() {
-		final List<RealLocalizableRealPositionable> pts = new ArrayList<>(3);
-		pts.add(new RealLocalizableRealPositionableWrapper<>(new RealPoint(
-			new double[] { 12, 3 })));
-		pts.add(new RealLocalizableRealPositionableWrapper<>(new RealPoint(
-			new double[] { 0.25, 6.5 })));
-		pts.add(new RealLocalizableRealPositionableWrapper<>(new RealPoint(
-			new double[] { -107, 33 })));
+		final List<RealPoint> pts = new ArrayList<>(3);
+		pts.add(new RealPoint(12, 3));
+		pts.add(new RealPoint(0.25, 6.5));
+		pts.add(new RealPoint(-107, 33));
 
 		rpc = new DefaultWritableRealPointCollection<>(pts);
-		wrap = new RealPointCollectionWrapper(rpc);
+		wrap = new RealPointCollectionWrapper<>(rpc, () -> new RealPoint(2));
 
 		// NB: can't remove points without associated image
 		final ImagePlus i = IJ.createImage("Ramp", "8-bit ramp", 128, 128, 1);
@@ -138,6 +140,39 @@ public class RealPointCollectionWrapperTest {
 		assertTrue(pointsEqual());
 	}
 
+	/**
+	 * Ensures that {@link WritableRealPointCollection}s can wrap
+	 * {@link RealRandomAccess}es (i.e. something that is not a {@link RealPoint})
+	 */
+	@Test
+	public void testRPCOfRandomAccesses() {
+		RealRandomAccessible<UnsignedByteType> testImg = Views.interpolate(ArrayImgs
+			.unsignedBytes(10, 10), new NearestNeighborInterpolatorFactory<>());
+		RealRandomAccess<UnsignedByteType> ra1 = testImg.realRandomAccess();
+		ra1.setPosition(new int[] { 1, 1 });
+		WritableRealPointCollection<RealRandomAccess<UnsignedByteType>> rpc =
+			new DefaultWritableRealPointCollection<>(Collections.singletonList(ra1));
+
+		RealPointCollectionWrapper<RealRandomAccess<UnsignedByteType>> wrapped =
+			new RealPointCollectionWrapper<>(rpc, testImg::realRandomAccess);
+
+		// add a new point
+		wrapped.addPoint(2.0, 3.0);
+		wrapped.synchronize();
+
+		// assert there are now two RRAs
+		assertEquals(2, rpc.size());
+		Iterator<RealRandomAccess<UnsignedByteType>> iw = rpc.points().iterator();
+		RealRandomAccess<UnsignedByteType> ra = iw.next();
+		assertArrayEquals(new double[] { 1.0, 1.0 }, ra.positionAsDoubleArray(),
+			1e-6);
+		ra = iw.next();
+		assertArrayEquals(new double[] { 2.0, 3.0 }, ra.positionAsDoubleArray(),
+			1e-6);
+		assertFalse(iw.hasNext());
+
+	}
+
 	// -- Helper methods --
 
 	private boolean pointsEqual() {
@@ -146,8 +181,7 @@ public class RealPointCollectionWrapperTest {
 		final float[] yp = wrap.getFloatPolygon().ypoints;
 
 		int count = 0;
-		final Iterator<RealLocalizableRealPositionable> itr = rpc.points()
-			.iterator();
+		final Iterator<RealPoint> itr = rpc.points().iterator();
 		while (itr.hasNext()) {
 			final RealLocalizable pt = itr.next();
 			boolean match = false;

